### PR TITLE
call webhook when mxjob start/finish

### PR DESCRIPTION
**Pull request by killwing**
- call webhook when mxjob start/finish
  - adopt ClientManager from k8s.io/apiserver/pkg/admission/plugin/webhook/config.ClientManager
  - call hook func async with exponential backoff
- type rename: WebHook -> Webhook
- add vendor github.com/pborman/uuid

### CODE REVIEW

File: cmd/mxnet-operator/main.go [hunk 1]
```diff
@@ -32,6 +32,10 @@ func main() {
 	logrus.Infof("Watching %s, %s, %s, %d", apiVersion, kind, namespace, resyncPeriod)
 	sdk.Watch(apiVersion, kind, namespace, resyncPeriod)
 
-	sdk.Handle(stub.NewHandler())
+	handler, e := stub.NewHandler()
+	if e != nil {
+		logrus.Fatalf("Failed to create handler: %v", err)
+	}
+	sdk.Handle(handler)
```
Comment (houz42) [inline cmd/mxnet-operator/main.go:39]: Is it Ok if handler is nil ?

File: pkg/mx/handler.go [hunk 2]
```diff
@@ -584,6 +613,80 @@ func (h *Handler) satisfiedExpectations(mxjob *v1alpha1.MXJob) bool {
 	return satisfied
 }
 
+func (h *Handler) callHookAsync(ctx context.Context, mxjob *v1alpha1.MXJob, startOrFinish bool) {
```
Comment (houz42) [inline pkg/mx/handler.go:616]: 1. use an enum of `startOrFinish` may be better than bool, if there are other webhooks
2. it seems like `StartWebhook` and `FinishWebhook` are actually PostStartWebhook and PostStopWebhook, is there any chance we will need PreStartWebhook or PreStopWebhook?

File: pkg/mx/webhook_client_manager.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+// Originate from k8s.io/apiserver/pkg/admission/plugin/webhook/config.ClientManager
+type ClientManager struct {
+	authInfoResolver     webhook.AuthenticationInfoResolver
+	serviceResolver      webhook.ServiceResolver
+	negotiatedSerializer runtime.NegotiatedSerializer
+	cache                *lru.Cache
+}
+
+// NewClientManager creates a ClientManager.
+func NewClientManager() (ClientManager, error) {
```
Comment (houz42) [inline pkg/mx/webhook_client_manager.go:38]: Why not return *ClientManager?

File: cmd/mxnet-operator/main.go (hunk 1 above)
Comment (killwing) [reply cmd/mxnet-operator/main.go:39]: Fatalf will exit process

File: pkg/mx/webhook_client_manager.go (hunk 3 above)
Comment (killwing) [reply pkg/mx/webhook_client_manager.go:38]: just keep the original design, may refine later

File: pkg/mx/handler.go (hunk 2 above)
Comment (killwing) [reply pkg/mx/handler.go:616]: 1. will refact if there are other webhooks added
1. To be exact, now StartWebhook is called when pods of job are start to initialize, FinishWebhook is called when pods of job are start to terminate. Will revisit if we need more exact hooks.

**Review by houz42 — APPROVED** [review]
lgtm

**Pull request merged**